### PR TITLE
Support bitbucket Dir()

### PR DESCRIPTION
**Pull request by mzampetakis**
Tested by exposing the function through API.
Here are some example responses of the Dir() function:
```
[
    {
        "Name": "test",
        "Data": null
    },
    {
        "Name": ".gitignore",
        "Data": "IyBUaGVzZSBhcmUgc29tZSBleGFtcGxlcyBvZiBjb21tb25seSBpZ25vcmVkIGZpbGUgcGF0dGVybnMuCiMgWW91IHNob3VsZCBjdXN0b21pemUgdGhpcyBsaXN0IGFzIGFwcGxpY2FibGUgdG8geW91ciBwcm9qZWN0LgojIExlYXJuIG1vcmUgYWJvdXQgLmdpdGlnbm9yZToKIyAgICAgaHR0cHM6Ly93d3cuYXRsYXNzaWFuLmNvbS9naXQvdHV0b3JpYWxzL3NhdmluZy1jaGFuZ2VzL2dpdGlnbm9yZQoKIyBOb2RlIGFydGlmYWN0IGZpbGVzCm5vZGVfbW9kdWxlcy8KZGlzdC8KCiMgQ29tcGlsZWQgSmF2YSBjbGFzcyBmaWxlcwoqLmNsYXNzCgojIENvbXBpbGVkIFB5dGhvbiBieXRlY29kZQoqLnB5W2NvZF0KCiMgTG9nIGZpbGVzCioubG9nCgojIFBhY2thZ2UgZmlsZXMKKi5qYXIKCiMgTWF2ZW4KdGFyZ2V0LwpkaXN0LwoKIyBKZXRCcmFpbnMgSURFCi5pZGVhLwoKIyBVbml0IHRlc3QgcmVwb3J0cwpURVNUKi54bWwKCiMgR2VuZXJhdGVkIGJ5IE1hY09TCi5EU19TdG9yZQoKIyBHZW5lcmF0ZWQgYnkgV2luZG93cwpUaHVtYnMuZGIKCiMgQXBwbGljYXRpb25zCiouYXBwCiouZXhlCioud2FyCgojIExhcmdlIG1lZGlhIGZpbGVzCioubXA0CioudGlmZgoqLmF2aQoqLmZsdgoqLm1vdgoqLndtdgoK"
    },
    {
        "Name": "README.md",
        "Data": "IyBSRUFETUUgIwoKVGhpcyBSRUFETUUgd291bGQgbm9ybWFsbHkgZG9jdW1lbnQgd2hhdGV2ZXIgc3RlcHMgYXJlIG5lY2Vzc2FyeSB0byBnZXQgeW91ciBhcHBsaWNhdGlvbiB1cCBhbmQgcnVubmluZy4KCiMjIyBXaGF0IGlzIHRoaXMgcmVwb3NpdG9yeSBmb3I/ICMjIwoKKiBRdWljayBzdW1tYXJ5CiogVmVyc2lvbgoqIFtMZWFybiBNYXJrZG93bl0oaHR0cHM6Ly9iaXRidWNrZXQub3JnL3R1dG9yaWFscy9tYXJrZG93bmRlbW8pCgojIyMgSG93IGRvIEkgZ2V0IHNldCB1cD8gIyMjCgoqIFN1bW1hcnkgb2Ygc2V0IHVwCiogQ29uZmlndXJhdGlvbgoqIERlcGVuZGVuY2llcwoqIERhdGFiYXNlIGNvbmZpZ3VyYXRpb24KKiBIb3cgdG8gcnVuIHRlc3RzCiogRGVwbG95bWVudCBpbnN0cnVjdGlvbnMKCiMjIyBDb250cmlidXRpb24gZ3VpZGVsaW5lcyAjIyMKCiogV3JpdGluZyB0ZXN0cwoqIENvZGUgcmV2aWV3CiogT3RoZXIgZ3VpZGVsaW5lcwoKIyMjIFdobyBkbyBJIHRhbGsgdG8/ICMjIwoKKiBSZXBvIG93bmVyIG9yIGFkbWluCiogT3RoZXIgY29tbXVuaXR5IG9yIHRlYW0gY29udGFjdA=="
    }
]
```

And a response with more that 10 results (default bitbucket's pagination):
```
[
    {
        "Name": "retest",
        "Data": null
    },
    {
        "Name": "file",
        "Data": "Zmlyc3QgZmlsZQ=="
    },
    {
        "Name": "file1",
        "Data": ""
    },
    {
        "Name": "file10",
        "Data": ""
    },
    {
        "Name": "file2",
        "Data": ""
    },
    {
        "Name": "file3",
        "Data": ""
    },
    {
        "Name": "file4",
        "Data": ""
    },
    {
        "Name": "file5",
        "Data": ""
    },
    {
        "Name": "file6",
        "Data": ""
    },
    {
        "Name": "file7",
        "Data": ""
    },
    {
        "Name": "file8",
        "Data": ""
    },
    {
        "Name": "file9",
        "Data": "bGFzdCBmaWxl"
    }
]
```

### CODE REVIEW

**Comment by 6543** [comment]
@mzampetakis this pull is conflicting :(

and could you also add some tests?

If you have questions just tell

**Comment by 6543** [comment]
could you also update the docs (https://woodpecker-ci.org/docs/administration/forges/overview#supported-features) , as this will enable `Multiple workflows` support

**Comment by mzampetakis** [comment]
Thanks @6543 ! Will definitely do the test and update the docs. 
Please let me know if it's more appropriate to close this PR or convert it to WIP.

**Comment by 6543** [comment]
just push to it and tell if it's ready - we do sqush-merge ;)

File: server/forge/bitbucket/bitbucket.go [hunk 1]
```diff
@@ -223,8 +223,53 @@ func (c *config) File(ctx context.Context, u *model.User, r *model.Repo, p *mode
...
+// Dir fetches a folder from the bitbucket repository
+func (c *config) Dir(ctx context.Context, u *model.User, r *model.Repo, p *model.Pipeline, f string) ([]*forge_types.FileMeta, error) {
+	var page *string
+	repoPathFiles := []*forge_types.FileMeta{}
+	for {
+		filesResp, err := c.newClient(ctx, u).GetRepoFiles(r.Owner, r.Name, p.Commit, f, page)
+		if err != nil {
+			return nil, err
+		}
+		totalFiles := 0
```
Comment (anbraten) [inline server/forge/bitbucket/bitbucket.go:235]:
```suggestion
```

not used

File: server/forge/bitbucket/bitbucket.go [hunk 2]
```diff
@@ -223,8 +223,53 @@ func (c *config) File(ctx context.Context, u *model.User, r *model.Repo, p *mode
...
 }
 
-func (c *config) Dir(_ context.Context, _ *model.User, _ *model.Repo, _ *model.Pipeline, _ string) ([]*forge_types.FileMeta, error) {
-	return nil, forge_types.ErrNotImplemented
+// Dir fetches a folder from the bitbucket repository
+func (c *config) Dir(ctx context.Context, u *model.User, r *model.Repo, p *model.Pipeline, f string) ([]*forge_types.FileMeta, error) {
+	var page *string
+	repoPathFiles := []*forge_types.FileMeta{}
+	for {
+		filesResp, err := c.newClient(ctx, u).GetRepoFiles(r.Owner, r.Name, p.Commit, f, page)
```
Comment (anbraten) [inline server/forge/bitbucket/bitbucket.go:231]:
```suggestion
	repoPathFiles := []*forge_types.FileMeta{}
	client := c.newClient(ctx, u)
	for {
		filesResp, err := client.GetRepoFiles(r.Owner, r.Name, p.Commit, f, page)
```

Reuse client, also for `.FindSource`

File: server/forge/bitbucket/bitbucket.go [hunk 3]
```diff
@@ -223,8 +223,53 @@ func (c *config) File(ctx context.Context, u *model.User, r *model.Repo, p *mode
...
+				}
+			}
+			repoPathFiles = append(repoPathFiles, &repoFile)
+		}
+
+		// Check for more results page
+		if filesResp.Next == nil {
+			break
+		}
+		myUrl, err := url.Parse(*filesResp.Next)
```
Comment (anbraten) [inline server/forge/bitbucket/bitbucket.go:258]:
```suggestion
		nextPageUrl, err := url.Parse(*filesResp.Next)
```

to be a bit more explaining

**Comment by 6543** [comment]
why did you delete the tests via forcepush ?

**Comment by mzampetakis** [comment]
> why did you delete the tests via forcepush ?
> 
> unintentional or not ready?

Pushed by mistake. They are almost ready but had not the time to double check. Will complete this soon as well as the suggestions.
Thanks!

File: server/forge/bitbucket/bitbucket.go [hunk 4]
```diff
@@ -18,10 +18,10 @@ package bitbucket
 import (
 	"context"
 	"fmt"
+	"golang.org/x/oauth2"
```
Comment (6543) [inline server/forge/bitbucket/bitbucket.go:21]: imports got unsorted ... I'll push a fix tomorow ... or you are fist wit the fix ;)

**Review by anbraten — APPROVED** [review]

**Pull request merged**